internal/util/concurrency/pool.go

```diff
@@ -0,0 +1,45 @@
+package concurrency
+
+import "github.com/panjf2000/ants/v2"
+
+type Pool struct {
+	inner *ants.Pool
+}
+
+func NewPool(cap int, opts ...ants.Option) (*Pool, error) {
+	pool, err := ants.NewPool(cap, opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Pool{
+		inner: pool,
+	}, nil
+}
+
+func (pool *Pool) Submit(method func() (interface{}, error)) *Future {
+	future := newFuture()
+	err := pool.inner.Submit(func() {
+		defer close(future.ch)
+		res, err := method()
+		if err != nil {
+			future.err = err
+		} else {
+			future.value = res
+		}
+	})
+	if err != nil {
+		future.err = err
+		close(future.ch)
```

Comment:
Looks like double closing the channel here.

Reply:
The close() is called at line 23 only if no error with Submit()

Reply:
> The close() is called at line 23 only if no error with Submit()

@yah01 at line 23, `close (future.ch)` is called in a deferred call, which means `future.ch` is closed no matter there is an error or not. So at line 33, the channel is closed the second time if there was an error in the method. This might be verified if there is a unit test with method returns an error.

Reply:
@yah01 oh, my mistake. The error at line 31 is return by the pool not the method.